### ISSUE TITLE
Question: Best way to force reevaluate of ThemeResource binding

### ISSUE THREAD

**Issue by marcelwgn** [microsoft/microsoft-ui-xaml#5457]
As the title says, what is the best way to force a reevaluate of ThemeResource bindings? Switching theme does work however it is not ideal.

**Comment by StephenLPeters** [comment]
What is the scenario you are trying to solve?  Under what conditions would the value you are bound to change that isn't picked up by ThemeResource automatically?

**Comment by chingucoding** [comment]
I was adding a ResourceDictionary manually to the Merged dictionaries of a control. However the resources did not get picked up, switching the RequestedTheme results in the resources bein re-evaluated though and the resources from the newly added ResourceDictionary were used.

**Comment by StephenLPeters** [comment]
I don't think we have anything that can instruct a ThemeResource reference to reevaluate... The other option I think would work would be to re-apply the template to the control.

**Comment by StephenLPeters** [comment]
@MikeHillberg maybe you know some magic api?

**Comment by MikeHillberg** [comment]
The only trick I know of is to change the RequestedTheme (and then change it back so no net effect). Since the change is all within one tick it doesn't cause flashing, it just causes the ThemeResource references to get re-evaluated.

**Comment by asklar** [comment]
See #4443

**Comment by chingucoding** [comment]
I assume a feature proposal/request for this would likely not be implemented would it? Could  we create some form of mechanism such that controls know that a dictionary affecting this control has changed? Though this would probably be too expensive on CPU cycles compared to the benefit it would create.

**Comment by MikeHillberg** [comment]
I think it would be reasonable to have some kind of feature to enable this. The current APIs and behavior reflect a lot of concern and hesitancy in the team around perf. {StaticResource} is great for perf because it's load time only. {DynamicResource} in WPF is great for dynamism, but there's working set and code complexity overhead. {ThemeResource} in WinUI wasn't trying to be as dynamic as {DynamicResource}, but then there's this RequestedTheme trick.

{x:Bind} is another way to get dynamism, and can be an alternative depending on the scenario.

**Comment by chingucoding** [comment]
How would you/we approach this? Should this automatically work or should we expose a function that would force reevaluating theme resourcebinding for a control and the whole VisualTree contained with it?

**Comment by MikeHillberg** [comment]
It might be as simple as adding a method that does the equivalent of poking the RequestedTheme property. I think ThemeResource is getting a big perf gain over DynamicResource today because the latter needs to re-evaluate everything every time a dependency _might_ need it. But the details are interesting enough that I'd like to see a spec first with more detail, including to revisit the scenarios and make sure we understand them.

**Comment by chingucoding** [comment]
Seems reasonable, especially since the performance overhead of that function wouldn't be to high. Would you like me to convert this issue into a more thought out proposal?

**Comment by MikeHillberg** [comment]
I'd like to make this part of a wider-reaching look at Xaml resources and bring some more people in. Multiple people have looked at related areas where we could make improvements (for example the one that asklar linked). For now though I don't think we have the time to focus on it yet.
